Source/items.cpp

```diff
@@ -3292,6 +3292,87 @@ Item *SpawnUnique(_unique_items uid, Point position, std::optional<int> level /*
+
+	// Gather all potential unique items. uid is the index into UniqueItems.
+	for (size_t i = 0; i < UniqueItems.size(); ++i) {
+		const UniqueItem &uniqueItem = UniqueItems[i];
+		// Verify the unique item base item matches idx.
+		bool isMatchingItemId = uniqueItem.UIItemId == AllItemsList[static_cast<size_t>(idx)].iItemId;
+		// Verify itemLvl is at least the unique's minimum required level.
+		// mLevel remains unadjusted when arriving in this function, and there is no call to set iblvl until CheckUnique(), so we adjust here.
+		bool meetsLevelRequirement = ((uper == 15) ? mLevel + 4 : mLevel) >= uniqueItem.UIMinLvl;
+		// Verify item hasn't been dropped yet. We set this to true in MP, since uniques previously dropping shouldn't prevent further identical uniques from dropping.
+		bool uniqueNotDroppedAlready = !UniqueItemFlags[i] || gbIsMultiplayer;
+
+		int uid = static_cast<int>(i);
+		if (IsUniqueAvailable(uid) && isMatchingItemId && meetsLevelRequirement && uniqueNotDroppedAlready)
+			uids.emplace_back(uid);
+	}
+
+	// If we find at least one unique in uids that hasn't been obtained yet, we can proceed getting a random unique.
+	if (uids.empty()) {
+		// Set uper to 1 and make the level adjustment so we have better odds of not generating a unique item.
+		if (uper == 15)
+			mLevel += 4;
+		uper = 1;
+
+		Point itemPos = item.position;
+
+		// Force generate a non-unique item.
+		DiabloGenerator itemGenerator(GetLCGEngineState());
+		do {
+			item = {}; // Reset item data
+			item.position = itemPos;
+			SetupAllItems(*MyPlayer, item, idx, itemGenerator.advanceRndSeed(), mLevel, uper, onlygood, pregen);
+		} while (item._iMagical == ITEM_QUALITY_UNIQUE);
+
+		return;
+	}
+
+	int32_t uidsIdx = std::max<int32_t>(0, GenerateRnd(static_cast<int32_t>(uids.size()))); // Index into uids, used to get a random uid from the uids vector.
+	int uid = uids[uidsIdx];                                                                // Actual unique id.
+
+	const UniqueItem &uniqueItem = UniqueItems[uid];
+	int targetLvl = 1; // Target level for reverse compatibility, since vanilla always takes the last applicable uid in the list.
+
+	// Set target level. Ideally we use uper 15 to have a 16% chance of generating a unique item.
+	if (uniqueItem.UIMinLvl - 4 > 0) { // Negative level will underflow. Lvl 0 items may have unintended consequences.
+		uper = 15;
+		targetLvl = uniqueItem.UIMinLvl - 4;
+	} else {
+		uper = 1;
+		targetLvl = uniqueItem.UIMinLvl;
+	}
+
+	// Amount to decrease the final uid by in CheckUnique() to get the desired unique.
+	const int uidOffset = std::count_if(UniqueItems.begin() + uid + 1, UniqueItems.end(), [&uniqueItem](UniqueItem &potentialMatch) {
+		return uniqueItem.UIItemId == potentialMatch.UIItemId && uniqueItem.UIMinLvl == potentialMatch.UIMinLvl;
+	});
+
+	Point itemPos = item.position;
+
+	// Force generate items until we find a uid match.
+	DiabloGenerator itemGenerator(GetLCGEngineState());

```

Comment:
If this gets merged after #7030, we can use xoshiro to generate item seeds instead. Doing so would help avoid RNG overlap, since the LCG engine state at this point is going to depend on the seed the item originally had.

Also, it occurs to me that it might make more sense to use the item seed directly here instead of `GetLCGEngineState()`.